File: bindings/php/uhppote/decode.php
<?php

{{- range .model.responses}}
{{- template "response" . -}}
{{end}}

{{range .model.responses}}
{{- template "decode" . -}}
{{end}}

{{- with .model.event}}
{{- template "decode" . -}}
{{end}}

{{- define "response"}}
class {{CamelCase .name}}
{   {{range .fields}}
    public ${{snakeCase .name}};{{end}}
}
{{end}}

{{- with .model.event -}}
class {{CamelCase .name}}
{   {{range .fields}}
    public ${{snakeCase .name}};{{end}}
}
{{end}}

{{define "decode"}}
function {{snakeCase .name}}($packet)
{
    if (count($packet) != 64) {
        throw new Exception(sprintf('invalid reply packet length (%d)', count($packet)));
    }

    // Ref. v6.62 firmware event
    if ($packet[0] != 0x17 && ($packet[0] != 0x19 || $packet[1] != 0x20)) {
        throw new Exception(sprintf('invalid reply start of message byte (%02x)', $packet[0]));
    }

    if ($packet[1] != {{byte2hex .msgtype}}) {
        throw new Exception(sprintf('invalid reply function code (%02x)', $packet[1]));
    }

    $response = new {{CamelCase .name}}();
 
    {{range .fields}}
    $response->{{snakeCase .name}} = unpack_{{snakeCase .type}}($packet, {{.offset}});
    {{- end}}

    return $response;
}
{{end}}

function unpack_uint8($packet, $offset)
{
    return $packet[$offset] & 0x00ff;
}

function unpack_uint16($packet, $offset)
{
    $v = 0;
    $v |= $packet[$offset+1] & 0x00ff;
    $v <<= 8;
    $v |= $packet[$offset] & 0x00ff;

    return $v;
}

function unpack_uint32($packet, $offset)
{
    $v = 0;
    $v = $packet[$offset+3] & 0x00ff;
    $v <<= 8;
    $v |= $packet[$offset+2] & 0x00ff;
    $v <<= 8;
    $v |= $packet[$offset+1] & 0x00ff;
    $v <<= 8;
    $v |= $packet[$offset] & 0x00ff;

    return $v;
}

function unpack_bool($packet, $offset)
{
    if (($packet[$offset] & 0x00ff) == 1) {
        return true;
    } else {
        return false;
    }
}

function unpack_IPv4($packet, $offset)
{
    return sprintf(
        '%d.%d.%d.%d',
        $packet[$offset],
        $packet[$offset+1],
        $packet[$offset+2],
        $packet[$offset+3]
    );
}

function unpack_MAC($packet, $offset)
{
    return sprintf(
        '%02x:%02x:%02x:%02x:%02x:%02x',
        $packet[$offset],
        $packet[$offset+1],
        $packet[$offset+2],
        $packet[$offset+3],
        $packet[$offset+4],
        $packet[$offset+5]
    );
}

function unpack_version($packet, $offset)
{
    $major = $packet[$offset];
    $minor = $packet[$offset+1];

    return sprintf("v%x.%'.02x", $major, $minor);
}

function unpack_date($packet, $offset)
{
    $bcd = bcd2string(array_slice($packet, $offset, 4));
    $year = substr($bcd, 0, 4);
    $month = substr($bcd, 4, 2);
    $day = substr($bcd, 6, 2);

    return sprintf('%s-%s-%s', $year, $month, $day);
}

function unpack_shortdate($packet, $offset)
{
    $bcd = bcd2string(array_slice($packet, $offset, 3));
    $year = substr($bcd, 0, 2);
    $month = substr($bcd, 2, 2);
    $day = substr($bcd, 4, 2);

    return sprintf('20%s-%s-%s', $year, $month, $day);
}

function unpack_optional_date($packet, $offset)
{
    $bcd = bcd2string(array_slice($packet, $offset, 4));
    $year = substr($bcd, 0, 4);
    $month = substr($bcd, 4, 2);
    $day = substr($bcd, 6, 2);

    if ($bcd == '00000000000000') {
        return '';
    } else {
        return sprintf('%s-%s-%s', $year, $month, $day);
    }
}

function unpack_time($packet, $offset)
{
    $bcd = bcd2string(array_slice($packet, $offset, 3));
    $hour = substr($bcd, 0, 2);
    $minute = substr($bcd, 2, 2);
    $second = substr($bcd, 4, 2);

    return sprintf('%s:%s:%s', $hour, $minute, $second);
}

function unpack_hhmm($packet, $offset)
{
    $bcd = bcd2string(array_slice($packet, $offset, 2));
    $hour = substr($bcd, 0, 2);
    $minute = substr($bcd, 2, 2);

    return sprintf('%s:%s', $hour, $minute);
}

function unpack_datetime($packet, $offset)
{
    $bcd = bcd2string(array_slice($packet, $offset, 7));
    $year = substr($bcd, 0, 4);
    $month = substr($bcd, 4, 2);
    $day = substr($bcd, 6, 2);
    $hour = substr($bcd, 8, 2);
    $minute = substr($bcd, 10, 2);
    $second = substr($bcd, 12, 2);

    return sprintf('%s-%s-%s %s:%s:%s', $year, $month, $day, $hour, $minute, $second);
}

function unpack_optional_datetime($packet, $offset)
{
    $bcd = bcd2string(array_slice($packet, $offset, 7));
    $year = substr($bcd, 0, 4);
    $month = substr($bcd, 4, 2);
    $day = substr($bcd, 6, 2);
    $hour = substr($bcd, 8, 2);
    $minute = substr($bcd, 10, 2);
    $second = substr($bcd, 12, 2);

    if ($bcd == '00000000000000') {
        return '';
    } else {
        return sprintf('%s-%s-%s %s:%s:%s', $year, $month, $day, $hour, $minute, $second);        
    }
}

function unpack_pin($packet, $offset) 
{
    $v = 0;
    $v |= $packet[$offset+2] & 0x00ff;
    $v <<= 8;
    $v |= $packet[$offset+1] & 0x00ff;
    $v <<= 8;
    $v |= $packet[$offset] & 0x00ff;

    return $v;
}

function unpack_mode($packet, $offset)
{
    return $packet[$offset] & 0x00ff;
}

function unpack_event_type($packet, $offset)
{
    return $packet[$offset] & 0x00ff;
}

function unpack_direction($packet, $offset)
{
    return $packet[$offset] & 0x00ff;
}

function unpack_reason($packet, $offset)
{
    return $packet[$offset] & 0x00ff;
}

function bcd2string($bytes)
{
    return join('', array_map(fn ($v) => sprintf('%02x', $v), $bytes));
}

File: bindings/go/uhppote/decode.go
package uhppote

import (
    "encoding/binary"
    "encoding/hex"
    "fmt"
    "net/netip"
    "regexp"
    "time"
)

{{range .model.responses}}
{{- template "response" . -}}
{{end}}

{{with .model.event}}
type {{CamelCase .name}} struct { {{range .fields}}
    {{CamelCase .name}} {{template "type" .type}} `json:"{{kebabCase .name}}"`{{end}}
}
{{end}}

{{range .model.responses}}
{{- template "decode" . -}}
{{end}}

{{with .model.event}}{{template "decode" .}}
{{end}}

{{define "decode"}}
func {{camelCase .name}}(packet []byte) (response *{{CamelCase .name}}, err error) {
    if len(packet) != 64 {
        err = fmt.Errorf("invalid reply packet length (%v)", len(packet))
        return
    }

    // Ref. v6.62 firmware event
    if packet[0] != 0x17 && (packet[0] != 0x19 || packet[1] != 0x20) {
        err = fmt.Errorf("invalid reply start of message byte (%02x)", packet[0])
        return
    }

    if packet[1] != {{byte2hex .msgtype}} {
        err = fmt.Errorf("invalid reply function code (%02x)", packet[1])
        return
    }

    v := {{CamelCase .name}}{}
    {{range .fields}}
    if v.{{CamelCase .name}},err = unpack{{CamelCase .type}}(packet, {{.offset}}); err != nil {
        return
    }
    {{end}}
    response = &v

    return
}
{{end}}

func unpackUint8(packet []byte, offset uint8) (uint8, error) {
    return packet[offset], nil
}

func unpackUint16(packet []byte, offset uint8) (uint16, error) {
    return binary.LittleEndian.Uint16(packet[offset:offset+2]), nil
}

func unpackUint32(packet []byte, offset uint8) (uint32, error) {
    return binary.LittleEndian.Uint32(packet[offset:offset+4]), nil
}

func unpackBool(packet []byte, offset uint8) (bool, error) {
    return packet[offset] != 0x00, nil
}

func unpackIPv4(packet []byte, offset uint8) (netip.Addr, error) {
    addr,_ := netip.AddrFromSlice(packet[offset:offset+4])

    return addr, nil
}

func unpackMAC(packet []byte, offset uint8) (string, error) {
    mac := fmt.Sprintf("%02x:%02x:%02x:%02x:%02x:%02x", 
        packet[offset],
        packet[offset+1],
        packet[offset+2],
        packet[offset+3],
        packet[offset+4],
        packet[offset+5])

    return mac, nil
}

func unpackVersion(packet []byte, offset uint8) (string, error) {
    major := packet[offset]
    minor := packet[offset+1]

    return fmt.Sprintf("v%x.%02x", major, minor), nil
}

func unpackDate(packet []byte, offset uint8) (Date, error) {
    bcd := bcd2string(packet[offset:offset+4])

    if date, err := time.ParseInLocation("20060102", bcd, time.Local); err != nil {
        return Date(time.Time{}), err
    } else {
        return Date(date), nil
    }
}

func unpackShortdate(packet []byte, offset uint8) (Date, error) {
    bcd := "20" + bcd2string(packet[offset:offset+3])

    if date, err := time.ParseInLocation("20060102", bcd, time.Local); err != nil {
        return Date(time.Time{}), err
    } else {
        return Date(date), nil
    }
}

func unpackOptionalDate(packet []byte, offset uint8) (Date, error) {
    bcd := bcd2string(packet[offset:offset+4])

    if d, err := time.ParseInLocation("20060102", bcd, time.Local); err != nil {
        return Date{}, nil
    } else {
        return  Date(d), nil
    }
}

func unpackDatetime(packet []byte, offset uint8) (DateTime, error) {
    bcd := bcd2string(packet[offset:offset+7])

    if date, err := time.ParseInLocation("20060102150405", bcd, time.Local); err != nil {
        return DateTime(time.Time{}), err
    } else {
        return DateTime(date), nil
    }
}

func unpackOptionalDatetime(packet []byte, offset uint8) (DateTime, error) {
    bcd := bcd2string(packet[offset:offset+7])

    if date, err := time.ParseInLocation("20060102150405", bcd, time.Local); err != nil {
        return DateTime{}, err
    } else {
        return DateTime(date), nil
    }
}

func unpackTime(packet []byte, offset uint8) (Time, error) {
    bcd := bcd2string(packet[offset:offset+3])

    if t, err := time.ParseInLocation("150405", bcd, time.Local); err != nil {
        return Time(time.Time{}), err
    } else {
        return Time(t), nil
    }
}

func unpackHHmm(packet []byte, offset uint8) (HHmm, error) {
    bcd := bcd2string(packet[offset:offset+2])

    if t, err := time.ParseInLocation("1504", bcd, time.Local); err != nil {
        return HHmm(time.Time{}), nil
    } else {
        return HHmm(t), nil
    }
}

func unpackPin(packet []byte, offset uint8) (PIN, error) {
    b := []byte { packet[offset+0], packet[offset+1], packet[offset+2], 0x00 }
    v := binary.LittleEndian.Uint32(b)

    return PIN(v), nil
}

func bcd2string(bytes []byte) string {
    BCD := hex.EncodeToString(bytes)

    if matched,err := regexp.MatchString(`^[0-9]*$`, BCD); err != nil || !matched {
        panic(fmt.Sprintf("invalid BCD value (%v)", bytes))        
    } 

    return BCD
}

{{define "response"}}
type {{CamelCase .name}} struct { {{range .fields}}
    {{CamelCase .name}} {{template "type" .type}} `json:"{{kebabCase .name}}"`{{end}}
}
{{end}}

